Software applications events

franzaps opened the pull request:
[Rendered](https://github.com/franzaps/nips/blob/applications/82.md)

This event represents an application as used in [Zapstore](https://zapstore.dev). It is actively being used in a production setting for many months already.

This relates to NIP-51 (kind 30063), those artifact sets point to their application (this event, kind 32267).

Feedback welcome

franzaps committed:
Software applications event

vitorpamplona commented:
hum.. isn't it possible to merge this with kind 31990? Kind 31990 is insufficiently weirdly defined on NIP but most apps already have their entry due to https://nostrapp.link/

We could just add the additional tags needed by zap.store to that definition.

franzaps commented:
NIP-89 is for apps that handle nostr events, isn't it?  It's not at all what I'm defining here. In fact, 90%+ of the apps on zap.store do not handle nostr events.

It feels very contrived to try to fit this simple definition on the complexity of NIP-89.

vitorpamplona commented:
Yes, that's what NIP-89 needed it for, but those app definitions don't need to exist to solely respond to Nostr events. Actually, I believe most apps don't event that "response" field setup correctly.

DVMs also reuse the same kind to define their  applications. 

Reusing the kind doesnt mean you have to implement NIP-89 to support yours.

franzaps commented:
Are you saying I can keep my schema as-is and change 32267 to 31990? In such case I would have to change `content` which will break other NIP-89 clients.

vitorpamplona commented:
Hum.. yeah I missed the `content` discrepancy. 

@pablof7z, do you see a possibility of merging those two? 

Or are we going to have to define our apps in two places?

franzaps commented:
Do you know what is the rationale for putting stringified JSON in `content` (unless it has to be encrypted)? I've noticed the pattern in several places

vitorpamplona commented:
> Do you know what is the rationale for putting stringified JSON in content

Dump decisions. They probably just copied how kind 0 works :(

franzaps commented:
I also need data to be in tags for full text indexing purpose.

31990 seems built for something else which is fine: all apps can sign a 32267 and if they want nostr interop also sign a 31990.

I will keep using kind 32267 unless there's a better alternative to describe an application

DanConwayDev commented:
Love to see this and look forward to adding releases to nip34 clients.

Instead of referencing a nip34 announcement event in `repository` tag it should be referenced with an `a` tag so it will be indexed by relays. clients should look for an a tag with `30617` kind if repository is omitted.

The release event:
1. should be moved into this nip from the example section of nip51.
2. should be immutable and not a replaceable event. then trust attestations could be placed on the event  eg. validation that the checksums of a reproducible build match. 
3. should include an optional `["commit-id","<SHA1-commit-id-used-to-build-release>"]` tag

From a style point of view: I see most nips describe events using jsonc rather than a table of tags, which is easier to read. Is this now the standard?

lastly, the app event could support the maintainers tag like nip34 announcements  This means clients could optionally interpret app events by the pubkeys listed with identical `d` tags as referring to the same app. `["maintainers", "<other-recognized-maintainer>", ...]`

franzaps commented:
> Love to see this and look forward to adding releases to nip34 clients.

Looking forward too!

> Instead of referencing a nip34 announcement event in `repository` tag it should be referenced with an `a` tag so it will be indexed by relays. clients should look for an a tag with `30617` kind if repository is omitted.

Sounds great

> The release event:
> 
>     1. should be moved into this nip from the example section of nip51.

Sure, I can do that

>     2. should be immutable and not a replaceable event. then trust attestations could be placed on the event  eg. validation that the checksums of a reproducible build match.

We're probably talking about different things. The release set is a NIP-51 replaceable event (kind 30063) that is simply a pointer to a set of software releases (binaries). Trust attestations should be applied to NIP-94 (kind 1063) events because they carry a checksum of the binaries. Moreover, if a developer made a mistake in the file collection or a typo in the release notes, it's useful that they can update the event.

>     3. should include an optional `["commit-id","<SHA1-commit-id-used-to-build-release>"]` tag

Sure - in kind 1063s

> From a style point of view: I see most nips describe events using jsonc rather than a table of tags, which is easier to read. Is this now the standard?

I don't think we have a standard for whether to include tables or not. Other NIPs use bullet points.

> lastly, the app event could support the maintainers tag like nip34 announcements This means clients could optionally interpret app events by the pubkeys listed with identical `d` tags as referring to the same app. `["maintainers", "<other-recognized-maintainer>", ...]`

This is currently done via attribution `p`/`zap` tags. I think a better idea is to include maintainers in releases (not apps).

bezysoftware commented:
> > Do you know what is the rationale for putting stringified JSON in content
> 
> Dumb decisions. They probably just copied how kind 0 works :(

Sorry to go off topic, but what is so bad about putting stringified json in content? Why are tags better?

franzaps commented:
@bezysoftware I would ask what is _good_ about putting stringified JSON in `content`. It makes no sense to encode JSON inside JSON for no gain (unless it needs to be encrypted) since values of that object are strings anyway. Tags can be indexed.

DanConwayDev commented on 82.md:
```diff
@@ -0,0 +1,35 @@
+| `d`         | App identifier (preferably in reverse domain notation)   | yes      | com.example.app       |
+| `name`      | Name of the app                             | yes      | Example App           |
+| `description` | Short description of the app              | no       | An app that exemplifies    |
+| `repository`| Source code repository URL or [NIP-34](34.md) repository (if open source) | no       | https://github.com/example/android or `30617:e4ee4fceb9721ce4ee4fceb91c9711ca2a6c9a5ab:example-repo` |
```
```suggestion
| `repository`| Source code repository URL | no       | https://github.com/example/android |
| `a`| address pointer to [NIP-34](34.md) repository | no       | `30617:e4ee4fceb9721ce4ee4fceb91c9711ca2a6c9a5ab:example-repo` |
| `commit-id`| SHA1 commit id used to build the release  | no       | `42981750c3ab3bacaf242bf05a623ae1b299e0a1` |

```

DanConwayDev reviewed (changes requested):
here is a suggested update reflecting the comments you were receptive to

franzaps committed:
Update 82.md

Co-authored-by: DanConwayDev <114834599+DanConwayDev@users.noreply.github.com>

franzaps committed:
Delegate app description and visuals to a NIP-54 event

franzaps committed:
Note on NIP-89

franzaps commented:
I rewrote good part of this NIP.

As suggested by @NielLiesmons application description should be a wiki article. With that, developers also get access to better tools for updating information.

Presented the tags in `jsonc` example format as suggested by @DanConwayDev 

Lastly, made a clarification on NIP-89.

Events on relay.zap.store do not yet conform to this spec but will be migrated shortly.

franzaps committed:
Icon URL

franzaps committed:
Update kind 32267 to reflect latest usage

franzaps commented:
Updated the NIP to reflect the usage on Zapstore.

Since it has been used for months now and it's pretty stable, can we get this merged?

staab commented:
Is this in use by anyone other than zap.store?

franzaps commented:
@staab Not that I'm aware of.

Other than, maybe, @Giszmo ?

staab commented:
Normally NIPs are merged once there's some compatibility that needs to be ensured. It's good to have it documented in a draft, but update here when there are other apps using the standard

franzaps commented:
This data structure will probably end up in [NDK](https://github.com/relaystr/ndk) and used in [Yana](https://github.com/frnandu/yana) and probably in a relay-based community started by @NielLiesmons

franzaps committed:
Add summary

vitorpamplona commented:
This should be merged.

staab commented:
Are there 2+ clients?

vitorpamplona commented:
I am putting in on Amethyst right now. We can wait for the next release if we don't have any other client using this yet.

DanConwayDev commented:
I intend to with gitworkshop and ngit but haven't got around to it yet.

-------- Original Message --------
On 22/01/2025 15:05, Vitor Pamplona  wrote:

> I am putting in on Amethyst right now. We can wait for the next release if we don't have any other client using this yet.
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/nostr-protocol/nips/pull/1336#issuecomment-2607493301), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/A3MDZJ5MAFJ5P3A7QHHGAUL2L6XT5AVCNFSM6AAAAABKCZBYVKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMBXGQ4TGMZQGE).
> You are receiving this because you were mentioned.Message ID: ***@***.***>

staab commented:
> I am putting in on Amethyst right now.

That's good enough for me. Are you implementing as is?

vitorpamplona commented:
Just to display right now. I am not sure it makes sense to create it yet.

franzaps commented:
@kehiy Absolutely, I am using it already for the Zapstore package manager (works on MacOS and Linux): https://zapstore.dev/docs/cli/manage/ -- it will get way more packages in the coming months

v0l commented:
There is no reason for the release sets to be replaceable, can that be changed now?

v0l commented:
Built a cli tool in rust to do publishing, no dependencies, no shell commands, only works for `APK` atm: https://git.v0l.io/Kieran/nap

franzaps commented:
> There is no reason for the release sets to be replaceable, can that be changed now?

The first question would be what do we gain by changing it

v0l commented:
> > There is no reason for the release sets to be replaceable, can that be changed now?
> 
> The first question would be what do we gain by changing it

Cant be rugged, small benefit

franzaps commented:
> Break things while you can. If there is no need for a replaceable event, let's get rid of it and clean up the spec.
> 
> Also, does `30063` only keep the latest version? Or is the d-tag used as versions?

I keep events of kind 30063 around for every version. I only use the latest one in the UI (for now). Yes the d-tag is used for identifier and version (`com.greenart7c3.nostrsigner@v2.0.1`)

I recall asking developers to use `--overwrite-release` in zapstore-cli multiple times over the past few months (it replaces the kind 30063). It happened for different reasons; the point is they make mistakes and need it.

I know what will happen if we make this a regular event: they will end up asking me to delete the bad event from the relay so they don't have to bump and re-release

vitorpamplona commented:
Can you add more guidance to the `30063`? Right now it is just a line in the spec. This d-tag for instance should be defined so that other stores can use the same.

franzaps commented:
> Can you add more guidance to the `30063`? Right now it is just a line in the spec. This d-tag for instance should be defined so that other stores can use the same.

Actually I think it's there simply for convenience because `version` should be picked up from the linked `e` 1063 event and the identifier (`d`-tag) from the linked `a` 32267 event. So other stores could use another string (if I was doing that though, I would observe 30063 events in the wild first)

vitorpamplona commented:
It would be nice to have a common way of doing. I could imagine switching back one version and having to "search" for what each store/publisher is using as dtag.

v0l commented:
Also missing some documentation about the NIP-94, i events as per this PR but it doesnt open in Zapstore app,

What is `apk_signature_hash` for?

franzaps commented:
> Also missing some documentation about the NIP-94, i published events as per this PR but it doesnt open in Zapstore app,
> 
> What is `apk_signature_hash` for?

Should be `apk_certificate_hash` really (used in Zapstore to detect cert mismatches early which for sure will result in failed package updates). None of the NIP-94 extensions are documented. I don't know whether it should be another NIP or what.

vitorpamplona commented:
Pictures and Videos were on NIP-94 and recently moved away from it into their own event kinds. That seems to have been a good move. 

The same can happen here if software applications need their own NIP-94-like kind where not only specialized tags (like `apk_certificate_hash`) can be defined just for these file types but also the procedures to validate those hashes can be discussed.

ZigBalthazar commented:
+1 relay implementation:
https://github.com/dezh-tech/ddsr/tree/main/zapoli

fiatjaf committed:
NIP-77: Negentropy syncing (#1494)

Co-authored-by: Doug Hoyte <doug@hcsw.org>

mikedilger committed:
Add NIP-77 to the README (#1939)

franzaps committed:
New kind, WIP

dluvian reviewed (commented):
Software releases being replaceable have multiple downsides:
- Breaks chronological order: **Impossible** to figure out which is the latest release because the latest timestamp may be an edit. Querying 10 releases will not yield the 10 latest releases when an old release has been edited. 
- Can't differentiate between initial release and update: A Github-style activity feed may only want to display new releases and not their edits.
- Minor downside of getting rugged by edits.

I would prefer a non-replaceable release event and a replaceable edit event for fixing bad releases. I we keep the replaceability then edits should add an initial-time timestamp tag to keep the `created_at` of the initial release event. This tag aids in ordering releases and differentiating edit from initial release.

dluvian commented:
Releases are replaceable to fix typos and broken artifacts but you're only supposed to fix the latest release or otherwise you will break the release order. Or you use specialized software that hack around this design flaw.

> I have never seen this happen in practice

Yeah, but to be fair there are not many people that are publishing on nostr right now.

What about using 30063 as the edit event for a non-replaceable release event? Both with same structure described in your proposal. So we can have sane ordering with editable releases.

franzaps commented:
> Releases are replaceable to fix typos and broken artifacts but you're only supposed to fix the latest release or otherwise you will break the release order. Or you use specialized software that hack around this design flaw.

Does a similar mechanism exist for long-form articles (kind 30023)? Or any other replaceable for that matter? Why are we singling this one out?

If you are a developer who modifies a 6-month old release (unlikely but possible) to presumably improve correctness of your event, why wouldn't you also take care of the correctness of the `created_at` date? It's not like you are forced to timestamp it with the current date.

An app that wishes to show latest releases could also query and sort client side by `version` (or even `version_code` if Android).

With a little creativity you can work around it. Query for the latest assets (kind 3063) for a specific platform (`f` tag) and you will get non-replaceable events that have a `version` tag in them. Sort accordingly.

> What about using 30063 as the edit event for a non-replaceable release event? Both with same structure described in your proposal. So we can have sane ordering with editable releases.

While I agree your proposal is technically more correct, there is a tradeoff of one additional query **per release** just to check if there were any edits. For something low-risk and for which there are more than decent alternatives as I laid out. Tradeoffs balanced on leaving it unchanged in my opinion.

dluvian commented:
> Why are we singling this one out?

Because it's much more important to figure out the latest release than the latest blog post.

> While I agree your proposal is technically more correct, there is a tradeoff of one additional query per release just to check if there were any edits. For something low-risk and for which there are more than decent alternatives as I laid out. Tradeoffs in favor of leaving it unchanged in my opinion.

That's fair 🤝

dluvian commented on 51.md:
```diff
@@ -58,7 +58,6 @@ Aside from their main identifier, the `"d"` tag, sets can optionally have a `"ti
 | Kind mute sets        | 30007 | mute pubkeys by kinds<br>`"d"` tag MUST be the kind string                                   | `"p"` (pubkeys)                                                                     |
 | Interest sets         | 30015 | interest topics represented by a bunch of "hashtags"                                         | `"t"` (hashtags)                                                                    |
 | Emoji sets            | 30030 | categorized emoji groups                                                                     | `"emoji"` (see [NIP-30](30.md))                                                     |
-| Release artifact sets | 30063 | group of artifacts of a software release                                                     | `"e"` (kind:1063 [file metadata](94.md) events), `"a"` (software application event) |
```
Is this used by anyone?
30063 is already reserved in rust-nostr as `Kind::ReleaseArtifactSet` @yukibtc

franzaps replied:
You remind me that we should update NIP-51 with changes here (mention that kind 30063 is covered here, and that it should point to kind 3063 events)

franzaps committed:
Move commit-id to release, add r tag, note on release version, mark optional tags

franzaps committed:
Further clarifications on independence of identifiers and versions

franzaps committed:
Rename to commit

dluvian commented:
I'm already using `Software Release kind 30063` to show new software releases in the activity feed in [Gitplaza](https://codeberg.org/dluvian/gitplaza) and I would like to support listing and creating repo releases with the same kind. Can you add an optional `a` tag to the documentation so we can point releases to a repo? @franzaps

franzaps commented:
> I'm already using `Software Release kind 30063` to show new software releases in the activity feed in [Gitplaza](https://codeberg.org/dluvian/gitplaza) and I would like to support listing and creating repo releases with the same kind. Can you add an optional `a` tag to the documentation so we can point releases to a repo? @franzaps

@dluvian Very cool!

The tag is available in the application kind 32267 (not in 30063):

```json
    [
      // OPTIONAL
      // Pointer to NIP-34 repository
      "a", "30617:<destination-pubkey>:<repo-id>", "<relay-url>"
    ],
```

FYI it will take 6+ weeks for Zapstore to start the migration to this new format. There are currently too many signers and we need to do this carefully.

dluvian commented on 82.md:
```diff
@@ -0,0 +1,316 @@
+    [
+      // REQUIRED, MULTIPLE
+      // Asset ID
+      "e", "<asset-id>"
```
I'm having trouble finding the assets, because there is no pubkey hint for a relay list look up. Can you add it?

franzaps committed:
- Add title to releases
- Add relay hints to asset IDs
- Use url for release URL
- Use "platform" instead of "operating system"
- Rename to apk_certificate_hash

franzaps commented:
Just pushed an update @dluvian @basantagoswami 

- Add title to releases
- Add relay hints to asset IDs
- Use url for release URL
- Use "platform" instead of "operating system"
- Rename to apk_certificate_hash

Migration will start just after the next release with this version of the NIP

franzaps committed:
Move commit to asset for reproducible builds

franzaps committed:
Fix typo for app ID

franzaps committed:
Minimum allowed version (pinning)

franzaps committed:
Remove release title

franzaps committed:
Add min_allowed_version_code

franzaps committed:
Remove release URL (pull notes to content if needed)

franzaps committed:
Remove permissions - content analysis done elsewhere

franzaps committed:
Merge branch 'master' into applications

franzaps committed:
Simplify platform handling, add MIME types and version comparison, add many more platforms

franzaps committed:
Add web bundles (PWA)

franzaps committed:
Explanation for native executables
